systimerfactory/ipowercontrollersubscriber.cpp

```diff
@@ -0,0 +1,277 @@
+										   const PowerController_PowerState_t newState,
+										   void *userdata)
+{
+	RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]:Entering \n", __FUNCTION__, __LINE__);
+	IpowerControllerSubscriber* instance = IpowerControllerSubscriber::getInstance();
+
+    	if (instance)
+	{
+		std::lock_guard<std::mutex> lock(instance->m_pwrEvtQueueLock);
+		instance->m_pwrEvtQueue.emplace(currentState, newState);
+	}
+	RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]:Sending Signal to Thread for Processing Callback Event\n", __FUNCTION__, __LINE__);
+	instance->m_pwrEvtCondVar.notify_one();
+}
+
+
+void IpowerControllerSubscriber::sysTimeMgrPwrConnectHandlingThreadFunc()
+{
+	RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]: Entered \n", __FUNCTION__, __LINE__);
+	uint32_t retValPwrCtrl=0;
+
+	/* Loop and check for Power controller connection and if fails sleep and retry */
+	while(true)
+	{
+		if(POWER_CONTROLLER_ERROR_NONE == PowerController_Connect())
+		{
+			RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]: PowerController_Connect is Success\n", __FUNCTION__, __LINE__);
+			break;
+		}
+		else
+		{
+			RDK_LOG(RDK_LOG_ERROR, LOG_SYSTIME, "[%s:%d]:PowerController_Connect Failed\n", __FUNCTION__, __LINE__);
+			/* Sleep for 300 msec */
+			usleep(PWR_CNTRL_CONNECT_WAIT_TIME_MS);
+		}
+	}
+	retValPwrCtrl = PowerController_RegisterPowerModeChangedCallback(IpowerControllerSubscriber::sysTimeMgrPwrEventHandler, nullptr);
+	if(POWER_CONTROLLER_ERROR_NONE != retValPwrCtrl)
+	{
+		/* Ideally Call back should not fail and failure can happen only if reregister same function is done.
+		   Only Error is Logged here for Failure as this is a separate thread*/
+		RDK_LOG(RDK_LOG_ERROR, LOG_SYSTIME, "[%s:%d]:Power Controller RegisterPowerModeChangedCallback Failed retValPwrCtrl=[%d]\n", __FUNCTION__, __LINE__,retValPwrCtrl);
+	}
+	RDK_LOG(RDK_LOG_INFO,LOG_SYSTIME,"[%s:%d]:Power Controller Registration Thread Exits Gracefully retValPwrCtrl = %d\n",__FUNCTION__,__LINE__,retValPwrCtrl);
+}
+
+
+void IpowerControllerSubscriber::sysTimeMgrPwrEventHandlingThreadFunc()
+{
+	RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]: Entered \n", __FUNCTION__, __LINE__);
+
+	while (true)
+	{
+		std::unique_lock<std::mutex> lkVar(m_pwrEvtMutexLock);
+		RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]: Waiting for Events from Power Manager\n", __FUNCTION__, __LINE__); 
+		/* Wait until the queue is not empty or stop thread is signaled       */
+		m_pwrEvtCondVar.wait(lkVar, [this]
+		{
+			return !m_pwrEvtQueue.empty() || m_sysTimeMgrPwrStopThread;
+		});
+		if (m_sysTimeMgrPwrStopThread)
+		{
+			RDK_LOG(RDK_LOG_INFO, LOG_SYSTIME, "[%s:%d]: Exiting thread - stop flag set\n", __FUNCTION__, __LINE__);
+			lkVar.unlock();
+			break;
+		}
+		/* Unlock before processing event from the Queue */
+		lkVar.unlock();		
+		SysTimeMgr_Power_Event_State_t sysTimeMgrPwrEvent(POWER_STATE_UNKNOWN,POWER_STATE_UNKNOWN);
```

Comment:
This logic looks unneeded, can we directly move the queue item to SysTimeMgr_Power_Event_State_t?
we can change the scope of this variable to queue element processing. Also check to see if you can simplify locking and unlocking mechanism.